tests/test_elements/test_mol2mol/test_mol2mol_filter.py

```diff
@@ -14,68 +23,246 @@
+                SmartsFilter__filter_elements=new_input_as_list,
+            )
+            filtered_smiles_5 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_5, [SMILES_ANTIMONY, SMILES_METAL_AU])
+
+    def test_smarts_filter_parallel(self) -> None:
+        """Test if molecules are filtered correctly by allowed SMARTS patterns in parallel."""
+        smarts_pats: dict[str, IntOrIntCountRange] = {
+            "c": (4, None),
+            "Cl": 1,
+            "cc": (1, None),
+            "ccc": (1, None),
+            "cccc": (1, None),
+            "ccccc": (1, None),
+            "cccccc": (1, None),
+            "c1ccccc1": (1, None),
+            "cCl": 1,
+        }
+        smarts_filter = SmartsFilter(smarts_pats, mode="all", n_jobs=2)
+        pipeline = Pipeline(
+            [
+                ("Smiles2Mol", SmilesToMol()),
+                ("SmartsFilter", smarts_filter),
+                ("Mol2Smiles", MolToSmiles()),
+                ("ErrorFilter", ErrorFilter()),
+            ],
+        )
+        filtered_smiles = pipeline.fit_transform(SMILES_LIST)
+        self.assertEqual(filtered_smiles, [SMILES_CHLOROBENZENE])
+
+    def test_descriptor_filter(self) -> None:
+        """Test if molecules are filtered correctly by allowed descriptors."""
+        descriptors: dict[str, FloatCountRange] = {
+            "MolWt": (None, 190),
+            "NumHAcceptors": (2, 10),
+        }
+
+        descriptor_filter = RDKitDescriptorsFilter(descriptors)
+
+        pipeline = Pipeline(
+            [
+                ("Smiles2Mol", SmilesToMol()),
+                ("DescriptorsFilter", descriptor_filter),
+                ("Mol2Smiles", MolToSmiles()),
+                ("ErrorFilter", ErrorFilter()),
+            ],
+        )
+        filtered_smiles = pipeline.fit_transform(SMILES_LIST)
+        self.assertEqual(filtered_smiles, SMILES_LIST)
+
+        pipeline.set_params(DescriptorsFilter__mode="all")
+        filtered_smiles_2 = pipeline.fit_transform(SMILES_LIST)
+        self.assertEqual(filtered_smiles_2, [SMILES_CL_BR])
+
+        pipeline.set_params(DescriptorsFilter__keep_matches=False)
+        filtered_smiles_3 = pipeline.fit_transform(SMILES_LIST)
         self.assertEqual(
-            filtered_smiles, [SMILES_BENZENE, SMILES_CHLOROBENZENE, SMILES_CL_BR]
+            filtered_smiles_3,
+            [SMILES_ANTIMONY, SMILES_BENZENE, SMILES_CHLOROBENZENE, SMILES_METAL_AU],
         )
 
-    def test_invalidate_mixtures(self) -> None:
-        """Test if mixtures are correctly invalidated.
+        pipeline.set_params(DescriptorsFilter__mode="any")
+        filtered_smiles_4 = pipeline.fit_transform(SMILES_LIST)
+        self.assertEqual(filtered_smiles_4, [])
+
+        pipeline.set_params(
+            DescriptorsFilter__mode="any", DescriptorsFilter__keep_matches=True
+        )
 
-        Returns
-        -------
-        None
-        """
+        pipeline.set_params(
```

Comment:
Would it be possible to write as a loop?

Reply:
Yes, I implemented this - hope it is still clean enough